### DISCUSSION TITLE
Desktop AR possible?

### DISCUSSION THREAD

**Discussion by MikeFesta** [google/model-viewer#3301] [Q&A] [answered]
Is it possible to get the AR experience on a Windows or Mac computer with a webcam?  

Changing a setting in Chrome or Firefox or any browser to make it act like a mobile device would be ideal. I tried using developer tools in Chrome to simulate a mobile device, which does get the AR button to appear, however the console shows an error: 
model-viewer.js:54909 Request to present in WebXR denied. 
model-viewer.js:54910 DOMException: Failed to execute 'isSessionSupported' on 'XRSystem': Access to the feature "xr" is disallowed by permissions policy.
It then tries to fall back to Scene Viewer, which doesn't work because it's not Android.

If anyone knows of a workaround, simulator, or custom browser to get this to work, that would be awesome.

**Comment by elalish** [comment, marked as answer]
Not through WebXR. Maybe Snap has a library? What are you hoping to accomplish with MV and a webcam? Face try-on?

**Reply by MikeFesta** [reply]
Thanks for confirming. It's for a customer that want's to use a large touchscreen kiosk in their physical store to show products in AR in an empty part of the showroom. On their website they are using a viewer that I built with model-viewer and while they can still view and change the materials in 3D on the kiosk, they were hoping to also have the mobile AR experience.

It makes sense that it wouldn't work with just a webcam as there wouldn't be any accelerometer data or camera movement to help identify the ground plane. With a camera in a fixed position, I imagine that a custom app could be made to overlay the 3D model over the camera feed, but it sounds like that's not available out of the box with WebXR.

**Reply by elalish** [reply]
Indeed not. If you set the camera parameters to match the webcam's and disable `camera-controls`, you could probably overlay the transparent MV rendering on the camera feed and make it work. The only tricky part would be calibration, but that's inevitable.

**Reply by MikeFesta** [reply]
Yes, I was able to mock up a proof of concept using a video element with the webcam feed as the background and I think that it could do the trick. If the client moves forward, I'll build out a UI for calibration and an option to use a 360º photo from the location as the environmental lighting. I'll share a link to the final product if I end up making it.

<img width="502" alt="Screen Shot 2022-03-24 at 5 18 03 PM" src="https://user-images.githubusercontent.com/6376793/160012191-a5d3317c-27c1-44f5-bfc0-8044bf4bdda3.png">
